### ISSUE TITLE
Determine the root cause of hierarchical TruncatedNormal sampling failures

### ISSUE THREAD

**Issue by AlexanderFengler** [lnccbrown/HSSM#1282]
## Purpose

Determine why the rejected centered hierarchical `TruncatedNormal` candidate samples badly, and distinguish a PyMC correctness defect from difficult-but-correct posterior geometry, backend/adaptation behavior, prior calibration, initialization, and HSSM likelihood interactions.

This is now a **research tracker**, not a production release gate:

- PR #1277 is closed unmerged and its exact rejected candidate is frozen on `codex/1269-truncated-normal-experimental-base` at `9e7002f0`.
- PR #1288 is an independent research-only sink based on that frozen branch. It preserves the v2 harness and immutable smoke evidence and must not merge into `main` merely to retain the record.
- The general unmatched-group prior policy remains independent in #1272.
- The only future production intersection is a small evidence-selected default-policy PR based on merged #1272, with no experimental machinery.

## What the canonical smoke established

Canonical run [33410419447](https://github.com/lnccbrown/HSSM/actions/runs/33410419447) on exact commit `8fdd272b` rejected the automatic native centered-TN candidate:

- direct Gaussian PyMC lower-only and two-sided cells failed the frozen divergence screen;
- current Bambi and the HSSM softmax cell also failed;
- the LBA chain was pathological despite zero divergences: R-hat up to 2.97, bulk ESS 2.33, BFMI 0.029, and all retained draws at maximum tree depth;
- other TN smoke cells passed, so this is not universal TN failure;
- completed finite-difference and PyTensor/JAX checks passed at their evaluated starts; the LBA gradient diagnostic did not complete.

Therefore:

- the problem is not HSSM- or Bambi-only, because direct PyMC reproduces it;
- the evidence does **not** yet establish a PyMC density or gradient bug;
- the leading hypothesis is a centered hierarchical funnel sharpened by TN normalization/boundary coupling, with additional calibration/start/likelihood interactions;
- the linked-Normal controls are practical alternatives but change the statistical model, so they cannot establish the root cause.

Do not run the full frozen 720-cell v2 qualification for a candidate already rejected by smoke, and do not weaken its thresholds or reroll its seeds.

## Causal investigation

Preserve all v2 files/results byte-for-byte. Add a small direct-Gaussian v3 study with four representations of the **same natural-scale model**:

1. `native_centered`: PyMC native centered `TruncatedNormal` location and group effects.
2. `manual_centered`: independent TN log density in explicit transformed coordinates using `Flat` variables, a `Potential`, and all Jacobians.
3. `group_icdf_noncentered`: native TN location plus exact conditional inverse-CDF group latents.
4. `full_icdf_noncentered`: exact inverse-CDF non-centering for both the location and group effects.

Every representation retains:

```text
group_location ~ TruncatedNormal(base_mu, 0.25, lower, upper)
group_scale ~ Weibull(1.5, 0.3)
group_effect[g] ~ TruncatedNormal(group_location, group_scale, lower, upper)
y[g, i] ~ Normal(group_effect[g], 0.5)
```

### B1 — exact models and independent correctness oracle

Child branch: `codex/1282-tn-causal-oracle`, stacked on #1288.

- implement a NumPy/SciPy second-order jet oracle with no PyMC/PyTensor/JAX/model-builder imports;
- include the full transformed posterior: priors, Gaussian likelihood, TN normalization, support/positive transforms, and all log-Jacobians;
- compare value, gradient, and Hessian against PyTensor and JAX for lower-only and finite bounds, boundary-near/interior values, and scale regimes from 0.01 to 1.0;
- compile the manual and inverse-CDF graphs under both backend paths before freezing a sampling manifest.

### B2 — prospective causal sampling study

Freeze two exact known-failing Gaussian regimes:

- lower-only, outside anchor, weak information, float64;
- two-sided, near-boundary truth, float32.

Use identical natural data and natural starts across all representations/backends, then map those starts into each coordinate system. Use PyMC and NumPyro:

- smoke: 2 regimes × 4 representations × 2 backends × 1 dataset = 16 fits, 2 chains, 250 tune + 250 draws;
- confirmation: 2 × 4 × 2 × 5 datasets = 80 fits, 4 chains, 1,000 tune + 1,000 draws.

Run each four-representation comparison on one worker with fresh per-cell caches and counterbalanced order. Expected centered failures must not prevent the non-centered cells from running.

### B3 — evidence report

Commit only reviewed aggregates and a causal decision report. Raw chains remain workflow artifacts. The report must include:

- `root_cause_class`;
- `pymc_correctness_bug_supported`;
- `centered_failure_reproduced`;
- `same_model_nc_candidate`;
- `backend_specific`;
- `default_policy_implication`;
- `next_required_experiment`.

## Predeclared causal classifications

- `native-pymc-correctness-defect`: native value/gradient/Hessian disagrees with the independent oracle while manual centered agrees. Stop and produce a minimal upstream PyMC reproducer.
- `native-graph-lowering-or-adaptation`: both centered densities agree with the oracle, but only the native graph samples badly.
- `conditional-centering-root-cause`: both centered forms fail similarly; group-level inverse-CDF NC is healthy and posterior-equivalent.
- `location-and-conditional-centering-root-cause`: only full inverse-CDF NC is healthy.
- `backend-specific`: one representation consistently succeeds on only one backend.
- `tn-or-residual-geometry`: every exact TN representation fails despite oracle agreement.
- `mixed-inconclusive`: no predeclared pattern is met, including a case where shared natural starts make the centered models healthy. The next experiment then isolates initialization/calibration prospectively.

Healthy representations must meet the existing v2 convergence/geometry thresholds and agree on the natural-scale posterior (combined rank R-hat below 1.01 and posterior-mean differences within four combined MCSEs).

A successful inverse-CDF representation advances to HSSM likelihood isolation. It does **not** authorize a new production default without a separate HSSM qualification and product-policy PR.

## Historical context

- HSSM PR #246 had unexplained sampling failures under the old custom truncation wrapper; that is a warning, not this model.
- PyMC #4417 documented a historical wrong-gradient TN defect, but the completed current local probes have not reproduced it.
- Bambi #617 shows that a centered native hierarchy can work in some regimes.
- Bambi #1003 concerns dropped locations under Bambi's Normal non-centering and is a separate compatibility problem.

**Comment by AlexanderFengler** [comment]
Progress update from draft PR #1288:

- The underspecified v1 design has been replaced by a frozen executable v2 contract: 55 scenarios / 748 cells across smoke, qualification, and stress tiers.
- Result provenance now binds exact locked environment, source, observed PyTensor/JAX precision, data, transformed starts, and raw-chain bytes. The assessor derives paired efficiency, recovery, backend-agreement, and SBC decisions rather than trusting worker verdicts.
- Direct PyMC/Bambi and real HSSM geometry probes now cover recursive native `TruncatedNormal` graphs, linked-Normal controls, transformed starts, finite differences, native JAX autodiff, and HSSM likelihood parity. The complete focused suite passes (183 tests).
- The native JAX path requires `jax>=0.11.0` and `pytensor>=3.2.4`; those floors and both frozen environment locks are now part of the draft PR. TensorFlow Probability is deliberately not introduced.

This is not yet a numerical keep decision. The next commit is the artifact-producing sampling runner/workflow; after that we will run smoke first, inspect failures, and only then launch the predeclared qualification/stress evidence. #1277 remains draft and blocked until the final keep/fix/withdraw result.

**Comment by AlexanderFengler** [comment]
## Canonical smoke result: screening failure

Run [33410419447](https://github.com/lnccbrown/HSSM/actions/runs/33410419447) executed the frozen v2 smoke tier on exact source `8fdd272b` and semantic manifest SHA `05d8be96204f124abe723fb38f65080631d2cdcb7c9b776428201ee66045e15e`.

The assessor returned `screening-fail` and the workflow correctly skipped qualification and stress:

- 10 cells planned; 9 completed; 1 LBA diagnostics failure; 0 missing
- direct PyMC lower-only: 15/500 divergences (3.0%)
- direct PyMC two-sided near-boundary: 17/500 (3.4%)
- current Bambi lower-only: 5/500 (1.0%; the frozen rule is strictly `<1%`)
- HSSM softmax beta: 22/500 (4.4%)

Every completed finite-difference and PyTensor/JAX local gradient contract passed. This rules out the historical local wrong-gradient mechanism for those nine cells, but does not rule out high curvature away from the checked start.

The LBA cell's post-sampling diagnostics exception did not make its raw chain usable: R-hat reached 2.97, bulk ESS 2.33, BFMI 0.029, and all 500 retained draws used 1,023 leapfrog steps. Both chains started around natural `b=1.2` while generated group truths were 0.243–0.417, so default-start/likelihood interaction is part of that cell's pathology. Commit `dbb2858e` fixes only the hosted diagnostics binding error by compiling explicit ordered inputs and invoking them positionally; it does not reinterpret the failed chain.

### What this establishes

- The frozen centered native-`TruncatedNormal` candidate is a no-go as an automatic safe default in its current form.
- This is not purely an HSSM-likelihood problem because direct Gaussian PyMC cells failed.
- It is not solely the one-sided outside anchor because a two-sided in-support anchor also failed.
- It is not a universal statement that every `TruncatedNormal` hierarchy fails; several smoke cells passed.
- The gate and evidence machinery behaved as designed; thresholds and seeds will not be changed post hoc.

The smallest decision-grade attribution panel is the five frozen replicates for the direct lower-outside, lower-inside, and two-sided pairs plus HSSM LBA near/interior and softmax pairs. A location-aware non-centered hierarchy on an unconstrained support-link scale is also the leading prospective redesign, but it needs its own frozen candidate rather than being used to retroactively rescue v2.

#1277 remains draft and blocked.

**Comment by AlexanderFengler** [comment]
The first causal child is now open as draft PR #1294: https://github.com/lnccbrown/HSSM/pull/1294

B1 result: an independent NumPy/SciPy second-order oracle and four equivalent PyMC/PyTensor representations now agree in transformed log density, gradient, and Hessian for lower-only and finite bounds under both PyTensor and genuine JAX autodiff. The child remains research-only and performs no sampling.

Review found and fixed two defects in the experimental alternatives themselves before publication: ordinary-probability CDF underflow in remote tails, and an incorrect JAX subgradient/Hessian at the log-CDF branch point. Exact alpha=42 tail regressions now agree with SciPy and the oracle through second order.

This does not yet identify the smoke failures root cause or qualify a default. The next child will run matched data/seeds across native centered, manual centered, group-ICDF non-centered, and full-ICDF non-centered forms and apply the predeclared causal classifier.

**Comment by AlexanderFengler** [comment]
B1 is now complete as a full causal centering factorial in draft PR #1294.

The initial four-form panel omitted one necessary cell: inverse-CDF non-centering of the population location while retaining centered native TruncatedNormal group effects. Without that arm, B2 could not distinguish location-centering from a joint location/group-centering interaction.

Commit `15349023` adds the missing form. The final same-model panel is:

1. native centered location / centered groups;
2. manual centered location / centered groups (implementation control);
3. centered location / inverse-CDF-NC groups;
4. inverse-CDF-NC location / centered groups;
5. inverse-CDF-NC location / inverse-CDF-NC groups.

All five forms agree with the independent NumPy/SciPy oracle in transformed log density, gradient, and Hessian for the declared lower-only and finite-bound cases under PyTensor and genuine JAX autodiff. The exact focused suite is now 125 passing tests; Ruff, formatting, mypy, Pyrefly, and independent review also pass. Frozen v2 files and evidence remain unchanged.

B2 is prospectively frozen as a wholly additive v3 child:

- two exact v2-failing Gaussian regimes;
- both PyMC and NumPyro;
- immutable shared data and natural-scale starts across all forms;
- one indivisible five-form worker block for each regime/backend/dataset;
- 20 smoke fits;
- eight confirmation datasets, giving 32 blocks / 160 confirmation fits;
- classification within each regime only;
- unfavorable sampler outcomes are valid evidence, while missing oracle/artifact evidence blocks a decision.

Five datasets would only permit a descriptive replicated case series (`p=0.0625` for all five paired directions). Eight is the minimum declared count that can clear the familywise threshold for the five planned paired contrasts when the direction is consistent. This remains a focused causal experiment, not a production-default PR.
